File: web/src/lib/pathfind/diplomacy.ts
import type { PactState, TreatyState } from "../../types/wire";
import type { Leg, ProvinceOwners } from "./types";

function stance(
  mover: string,
  owner: string,
  diplomacy: TreatyState[] | undefined,
): "war" | "peace" | "alliance" {
  if (!owner || mover === owner) return "peace";
  const key = [mover, owner].sort().join("|");
  const t = diplomacy?.find(
    (d) => [d.slot_a, d.slot_b].sort().join("|") === key,
  );
  return (t?.stance as "war" | "peace" | "alliance") ?? "war";
}

function hasPact(
  from: string,
  to: string,
  kind: "right_of_way",
  pacts: PactState[] | undefined,
): boolean {
  return Boolean(pacts?.some((p) => p.from === from && p.to === to && p.kind === kind));
}

export function mayMoveThrough(
  mover: string,
  owner: string,
  diplomacy: TreatyState[] | undefined,
  pacts: PactState[] | undefined,
): boolean {
  if (!owner || mover === owner) return true;
  const s = stance(mover, owner, diplomacy);
  if (s === "alliance") return true;
  if (s === "peace") return hasPact(owner, mover, "right_of_way", pacts);
  return false;
}

export function hostileLegEnd(
  mover: string,
  owners: ProvinceOwners,
  leg: Leg,
  diplomacy: TreatyState[] | undefined,
): boolean {
  const owner = owners[leg.toProv] ?? "";
  if (!owner || owner === mover) return false;
  return stance(mover, owner, diplomacy) === "war";
}

export function routeBlocked(
  mover: string,
  owners: ProvinceOwners,
  legs: Leg[],
  diplomacy: TreatyState[] | undefined,
  pacts: PactState[] | undefined,
): boolean {
  for (const leg of legs) {
    const owner = owners[leg.toProv] ?? "";
    if (!owner || owner === mover) continue;
    if (mayMoveThrough(mover, owner, diplomacy, pacts)) continue;
    if (stance(mover, owner, diplomacy) !== "war") return true;
  }
  return false;
}

export function routeAttackTerminus(
  mover: string,
  owners: ProvinceOwners,
  legs: Leg[],
  diplomacy: TreatyState[] | undefined,
): boolean {
  return legs.some((l) => hostileLegEnd(mover, owners, l, diplomacy));
}

export function provinceOwners(
  provinces: { id: string; owner_id?: string }[],
): ProvinceOwners {
  const out: ProvinceOwners = {};
  for (const p of provinces) {
    out[p.id] = p.owner_id ?? "";
  }
  return out;
}

File: internal/domain/pathdom/diplomacy.go
package pathdom

import (
	"github.com/joaquing/clone-supremacy/internal/domain/diplomacydom"
)

// ProvinceOwners maps province id -> owner slot (empty = neutral).
type ProvinceOwners map[string]string

// LegBlocked reports whether mover may enter the leg's destination province.
func LegBlocked(d *diplomacydom.Registry, mover string, owners ProvinceOwners, leg Leg) bool {
	owner := owners[leg.ToProv]
	if owner == "" || owner == mover {
		return false
	}
	if d.MayMoveThrough(mover, owner) {
		return false
	}
	return d.Stance(mover, owner) != diplomacydom.War
}

// HostileLegEnd is true when the leg ends in enemy territory at war.
func HostileLegEnd(d *diplomacydom.Registry, mover string, owners ProvinceOwners, leg Leg) bool {
	owner := owners[leg.ToProv]
	if owner == "" || owner == mover {
		return false
	}
	return d.Stance(mover, owner) == diplomacydom.War
}

// RouteAttackTerminus is true if any leg ends in hostile territory.
func RouteAttackTerminus(d *diplomacydom.Registry, mover string, owners ProvinceOwners, legs []Leg) bool {
	for _, l := range legs {
		if HostileLegEnd(d, mover, owners, l) {
			return true
		}
	}
	return false
}

// RouteBlocked returns true if any leg is blocked by diplomacy (non-war).
func RouteBlocked(d *diplomacydom.Registry, mover string, owners ProvinceOwners, legs []Leg) bool {
	for _, l := range legs {
		if LegBlocked(d, mover, owners, l) {
			return true
		}
	}
	return false
}
